src/object_functions.c

```diff
@@ -1419,10 +1570,215 @@ void obj_init_bonus(Object *obj, LevelObjectEntry_Bonus *entry) {
+            func_800AFC3C(obj, speed);
+            obj->unk7C.word -= speed;
+        } else {
+            gParticlePtrList_addObject(obj);
+        }
+    } else {
+        obj->unk6 |= 0x4000;
+        if (obj->unk78 == 0) {
+            obj->unk6 &= 0xBFFF;
+            doubleSpeed = speed * 2;
+            if (obj->unk39 < (255 - doubleSpeed)) {
+                obj->unk39 += doubleSpeed;
+            } else {
+                obj->unk39 = 255;
+            }
+            obj4C = obj->unk4C;
+            if ((obj4C->unk13 < 45) && (someBool == FALSE)) {
+                racerObj = obj4C->unk0;
+                if (racerObj && (racerObj->header->behaviorId == 1) && (((Object_64_Racer *)racerObj->unk64)->unk0 == 0)) {
+                    settings->balloonsPtr[settings->worldId]++;
+                    if (settings->worldId != 0) {
+                        settings->balloonsPtr[0]++;
+                    }
+                    settings->courseFlagsPtr[settings->courseId] |=  flag;
+                    func_80001EA8(573, obj->x_position, obj->y_position, obj->z_position, NULL);
+                    obj->unk7C.word = 16;
+                    obj->unk74 = 2;
+                    obj->unk6 |= 0x4000;
+                    func_800AFC3C(obj, speed);
+                }
+            }
+            obj64 = (Object_64_GoldenBalloon *)obj->unk64;
+            obj->unk3B = 0;
+            obj64->unk14 = 0.0f;
+            if (obj->unk39 < 255) {
+                speedf = 0.0f;
+            } else {
+                speedf = 1.0f;
+            }
+            if (obj64->unkD == 255) {
+                obj64->unkD = func_8001C524(obj->x_position, obj->y_position, obj->z_position, 0);
+                if (obj64->unkD != 255) {
+                    obj64->unkE = func_8001CC48(obj64->unkD, -1, 0);
+                    obj64->unkF = func_8001CC48(obj64->unkE, obj64->unkD, 0);
+                    obj64->unk10 = func_8001CC48(obj64->unkF, obj64->unkE, 0);
+                    obj64->unkC = obj64->unkD;
+                }
+            } else {
+                func_8001C6C4(obj64, obj, sp2C, speedf, 0);
+            }
+        }
+    }
+}
+
+
+#else
 GLOBAL_ASM("asm/non_matchings/unknown_032760/obj_loop_goldenballoon.s")
-GLOBAL_ASM("asm/non_matchings/unknown_032760/obj_init_door.s")
+#endif
+
+void obj_init_door(Object *obj, LevelObjectEntry_Door *entry) {
+    Object_64_Door *obj64;
+    f32 phi_f0;
+
+    obj64 = (Object_64_Door *)obj->unk64;
+    if (entry->unkC == -1) {
+        entry->unkC = func_8000CC20(obj);
+    } else {
+        func_8000CBF0(obj, entry->unkC);
+    }
+    obj64->unkE = entry->unkC;
+    obj64->unkF = entry->unkE;
+    obj64->unk11 = entry->numBalloonsToOpen;
+    obj64->unk10 = entry->numBalloonsToOpen;
+    obj64->unk12 = entry->distanceToOpen;
+    if (obj64->unkE == -1) {
+        rmonPrintf("Illegal door no!!!\n");
+    }
+    obj->unk3A = entry->modelIndex;
+    obj->y_rotation = entry->closedRotation << 6 << 4;
```

Comment:
Just curious, since I've been seeing a lot of instances of `rotation << 6 << 4`. Assuming this isn't a fake match, I would guess that the bit shifts are split up because the inner shift is part of a macro expansion, e.g.
```
#define SHIFT_ROTATION(rotation) ((rotation) << 6)
obj->y_rotation = SHIFT_ROTATION(entry->closedRotation) << 4;
```
In that case, do you have any inkling of what that macro would represent?

To be clear, this isn't a suggestion to add a macro. I'm just curious if you know the purpose of this.

Reply:
I haven't got much of a clue to be honest, although I agree that it is probably a macro of some kind. I know it is a double-shift, because doing two shifts skips a temporary register in the assembly. It doesn't even have to be 6 and 4. It could be any two numbers that add up to 10. I just chose 6 & 4 arbitrarily I think, and just stuck with those numbers. 

In a LevelObjectEntry, rotation is usually only represented by a single byte. For most objects, like the doors, 1 unit is 1/64th of a circle. I think that is probably where `<< 6` could come in, and then the `<< 4` could be a conversion factor between the entry's rotation to the object's actual rotation value. If that makes any sense lol. 

There are some objects, like `lasergun`, that have the rotation byte unit value represented as 1/256th of circle.  In that function both shift values add up to 8 instead of 10. Need to look it more.

Reply:
Gotcha, yeah that makes sense. Thanks for humoring me.